pkg/controller/operators/olm/operatorgroup.go

```diff
@@ -532,11 +532,12 @@ func (a *Operator) ensureSingletonRBAC(operatorNamespace string, csv *v1alpha1.C
 			}
 			// TODO: this should do something smarter if the cluster role already exists
 			if cr, err := a.opClient.CreateClusterRole(clusterRole); err != nil {
+				if !k8serrors.IsAlreadyExists(err) {
+					return err
+				}
 				// if the CR already exists, but the label is correct, the cache is just behind
-				if k8serrors.IsAlreadyExists(err) && ownerutil.IsOwnedByLabel(cr, csv) {
+				if cr != nil && ownerutil.IsOwnedByLabel(cr, csv) {
 					continue
-				} else {
-					return err
 				}
```

Comment:
It looks like the @ecordell 's [original patch](https://github.com/operator-framework/operator-lifecycle-manager/commit/538b1315eace85e33df7b2893a85a000f8757349), effectively returns _any_ error if `!IsOwnedByLabel`, which this code does not. Assuming errors returned from this function trigger a resync -- i.e. a retry -- then we'll fail to retry the operation when we detect the cluster state isn't as expected.

I suspect something like the following will produce the desired results:

```go
if k8serrors.IsAlreadyExists(err) && cr != nil && ownerutil.IsOwnedByLabel(cr, csv) {
    continue
}
return err
```

Reply:
When I had played around with this exact implementation, we had reverted back to the original testing flake behavior (e.g. "crb-role already exists") as there's likely a larger issue in the fake testing client where the resource returned from the Create call is always going to be nil. With that said, it's likely we don't need to nil check the variable returned by Create as this is a testing client problem. Any opinions?

Reply:
Moving some of this from slack:

I had tried adding a client-go testing reactor for returning an object for these situations:

```diff
+// WithOperatorGroupReactors returns a fakeK8sClientOption that configures a Clientset to return the ClusterRole
+// or ClusterRoleBinding object during a create call.
+// Note(tflannag): Fix for https://github.com/operator-framework/operator-lifecycle-manager/issues/2091.
+func WithOperatorGroupReactors(tb testing.TB) Option {
+       return func(c ClientsetDecorator) {
+               c.PrependReactor("create", "clusterrolebindings", func(action clitesting.Action) (bool, runtime.Object, error) {
+                       createAction := action.(clitesting.CreateActionImpl)
+                       return false, createAction.Object, nil
+               })
+               c.PrependReactor("create", "clusterroles", func(action clitesting.Action) (bool, runtime.Object, error) {
+                       createAction := action.(clitesting.CreateActionImpl)
+                       return false, createAction.Object, nil
+               })
+       }
+}
```

But it wasn't immediately clear how to best integrate those reactors into the current fake client setup such that we're not always appending these reactors and increasing the time it takes to run unit tests, which likely indicates it's more useful as a one-off functional type setup:

```diff
-       k8sClientFake := k8sfake.NewSimpleClientset(config.k8sObjs...)
+       k8sClientFake := fake.NewClientSetDecoratorWithReactors(config.k8sObjs, config.fakeClientOptions...)
        k8sClientFake.Resources = apiResourcesForObjects(append(config.extObjs, config.regObjs...))
-       config.operatorClient = operatorclient.NewClient(k8sClientFake, apiextensionsfake.NewSimpleClientset(config.extObjs...), apiregistrationfake.NewSimpleClientset(config.regObjs...))
+       config.operatorClient = operatorclient.NewClient(k8sClientFake.Clientset, apiextensionsfake.NewSimpleClientset(config.extObjs...), apiregistrationfake.NewSimpleClientset(config.regObjs...))
```

I tried running down this route but got sidetracked with other work.

Reply:
> But it wasn't immediately clear how to best integrate those reactors into the current fake client setup such that we're not always appending these reactors and increasing the time it takes to run unit tests, which likely indicates it's more useful as a one-off functional type setup

imho this reactor is generally useful for all unit tests, but I'm not sure the error from the create action would be carried forward properly by the existing decorator even if the example you gave was added.

> When I had played around with this exact implementation, we had reverted back to the original testing flake behavior (e.g. "crb-role already exists")

this is expected in the implementation with a bad client fake, but it should no longer panic at test time, and _should_ do the correct thing at runtime; i.e. return an error -- even if it's not the "right" error -- when the cluster has a different view of the cluster.